packages/hash/api/src/model/knowledge/link.model.ts

```diff
@@ -111,19 +110,191 @@ export default class {
+        index,
         linkTypeId: linkTypeModel.schema.$id,
         targetEntityId: targetEntityModel.entityId,
       },
     );
 
-    return LinkModel.fromPersistedLink(graphApi, {
+    /** @todo: this should be returned directly from the `createLink` method */
+    const persistedLink = {
       inner: link,
       ownedById: createdBy,
+    };
+
+    return LinkModel.fromPersistedLink(graphApi, persistedLink);
+  }
+
+  /**
+   * Create a link between a source and a target entity using a specific link
+   * type.
+   *
+   * @param params.accountId the accountId of the account creating the link
+   * @param params.sourceEntityModel the source entity of the link
+   * @param params.linkTypeModel the Link Type of the link
+   * @param params.targetEntityModel the target entity of the link
+   */
+  static async create(
+    graphApi: GraphApi,
+    params: LinkModelCreateParams,
+  ): Promise<LinkModel> {
+    const { sourceEntityModel, linkTypeModel, createdBy } = params;
+    const siblingLinks = await sourceEntityModel.getOutgoingLinks(graphApi, {
+      linkTypeModel,
+    });
+
+    /** @todo: rely on Graph API validation instead of manually checking whether sibling links are ordered */
+    const hasOrderedSiblingLink = siblingLinks[0]?.index !== undefined;
+
+    const index = hasOrderedSiblingLink
+      ? // if siblings are ordered and an index is provided, use the provided index
+        params.index ??
+        // if siblings are ordered and no index is provided, default to the end of the list of links
+        siblingLinks.length
+      : siblingLinks.length === 0
+      ? // if siblings are not ordered because there are no siblings, allow the link to be ordered
+        params.index
+      : // if siblings are not ordered and there are siblings, don't allow the link to be ordered
+        undefined;
+
+    if (index !== undefined) {
+      /** @todo: rely on Graph API to validate the index */
+      if (index < 0 || index > siblingLinks.length) {
+        throw new Error("Provided link index is out of bounds");
+      }
+
+      await Promise.all(
+        siblingLinks
+          .filter((sibling) => sibling.index! >= index)
+          .map((sibling) =>
+            sibling.updateWithoutUpdatingSiblings(graphApi, {
+              updatedIndex: sibling.index! + 1,
+              updatedBy: createdBy,
+            }),
+          ),
+      );
+    }
+
+    return await LinkModel.createLinkWithoutUpdatingSiblings(graphApi, {
+      ...params,
+      index,
+    });
+  }
+
+  private async updateWithoutUpdatingSiblings(
+    graphApi: GraphApi,
+    params: { updatedIndex: number; updatedBy: string },
+  ) {
+    const { updatedIndex, updatedBy } = params;
+
+    const { index: previousIndex } = this;
+
+    if (previousIndex === undefined) {
+      throw new Error("Cannot make an un-ordered link ordered");
+    }
+
+    if (previousIndex === updatedIndex) {
+      throw new Error("No-op: link already has this index");
```

Comment:
Not suggesting this is wrong, but, why should this be an error?

Reply:
In prior workspace code we've been erroring on no-ops so that they're easier to catch. Agree with the thinking that erroring is strange behaviour when there's nothing necessarily wrong with it. Would you like me to remove it? In which case I'll reframe for throwing on no-ops in future.

Reply:
If this is consistent with the rest of the code, then perhaps it's fine to leave it and throw an asana task to revisit the pattern. I'm not necessarily against it, but it does seem quite strange yeah.

Reply:
[Asana task to track this ](https://app.asana.com/0/1202805690238892/1203055997033808/f)